### PR TITLE
add algorithm to check binary search tree

### PR DESCRIPTION
**Pull request by alexpantyukhin**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: data_structures/binary_tree/is_bst.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+from typing import Optional
+
+
+class TreeNode:
+    def __init__(
+        self,
+        data: float = 0,
+        left: Optional["TreeNode"] = None,
+        right: Optional["TreeNode"] = None,
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:33]:
```suggestion
        left: TreeNode | None = None,
        right: TreeNode | None = None,
```

File: data_structures/binary_tree/is_bst.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+Depth-first search approach.
+
+leetcode: https://leetcode.com/problems/validate-binary-search-tree/
+
+Let n is the number of nodes in tree
+Runtime: O(n)
+Space: O(1)
+"""
+
+from typing import Optional
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:25]: Python `3.10` introduced the pipe symbol as a union in typehints - You can just do `x | None`
```suggestion
```

File: data_structures/binary_tree/is_bst.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        return is_valid_tree(node.left) and is_valid_tree(node.right)
+
+    if not is_valid_tree(root):
+        raise ValueError(
+            "Each node should be type of TreeNode and data should be float."
+        )
+
+    def is_bst_internal(
+        node: Optional["TreeNode"], left_bound: float, right_bound: float
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:96]:
```suggestion
        node: TreeNode | None, left_bound: float, right_bound: float
```

File: data_structures/binary_tree/is_bst.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        if not isinstance(node, TreeNode):
+            return False
+
+        try:
+            float(node.data)
+        except ValueError:
+            return False
+        except TypeError:
+            return False
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:88]:
```suggestion
        except (ValueError, TypeError):
            return False
```

File: data_structures/binary_tree/is_bst.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        self.data = data
+        self.left = left
+        self.right = right
+
+
+def is_bst(root: TreeNode | None) -> bool:
+    """
+    >>> is_bst(TreeNode(2, TreeNode(1), TreeNode(3)))
+    True
+
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:44]:
```suggestion
```

File: data_structures/binary_tree/is_bst.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    >>> is_bst(TreeNode(0, TreeNode(-11), TreeNode(3)))
+    True
+
```
Comment (CaedenPH) [inline data_structures/binary_tree/is_bst.py:47]:
```suggestion
```

**Review by CaedenPH — COMMENTED** [review]
Other than these small points looks good

File: data_structures/binary_tree/is_bst.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        data: float = 0,
+        left: TreeNode | None = None,
+        right: TreeNode | None = None,
+    ) -> None:
+        self.data = data
+        self.left = left
+        self.right = right
+
+
+def is_bst(root: TreeNode | None) -> bool:
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:40]: For readability, we try to expand acronyms so `is_binary_search_tree()` or maybe even better...
```suggestion
def is_balanced(root: TreeNode | None) -> bool:
```

File: data_structures/binary_tree/is_bst.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        """
+        if node is None:
+            return True
+
+        if not isinstance(node, TreeNode):
+            return False
+
+        try:
+            float(node.data)
+        except (ValueError, TypeError):
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:83]:
```suggestion
        except (TypeError, ValueError):
```

File: data_structures/binary_tree/is_bst.py (hunk 7 above)
Comment (alexpantyukhin) [reply data_structures/binary_tree/is_bst.py]: I applied is_binary_search_tree. I think it's more close that the func does. If I correctly understand: balanced - heights keft right is close to each other

File: data_structures/binary_tree/is_bst.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            return False
+
+        return is_valid_tree(node.left) and is_valid_tree(node.right)
+
+    if not is_valid_tree(root):
+        raise ValueError(
+            "Each node should be type of TreeNode and data should be float."
+        )
+
+    def is_bst_internal(
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py]:
```suggestion
    def is_binary_search_tree_internal(
```
Needs a docstring that explains criteria.
Needs at least one doctest that returns `False`

File: data_structures/binary_tree/is_bst.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+"""
+Author  : Alexander Pantyukhin
+Date    : November 2, 2022
+
+Task:
+Given the root of a binary tree, determine if it is a valid binary search
+tree (BST).
+
+A valid BST is defined as follows:
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:9]:
```suggestion
A valid binary search tree is defined as follows:
```

File: data_structures/binary_tree/is_bst.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+- Both the left and right subtrees must also be binary search trees.
+
+Implementation notes:
+Depth-first search approach.
+
+leetcode: https://leetcode.com/problems/validate-binary-search-tree/
+
+Let n is the number of nodes in tree
+Runtime: O(n)
+Space: O(1)
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:22]: How do we _prove_ these statements?

File: data_structures/binary_tree/is_bst.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+"""
+
+from __future__ import annotations
+
+
+class TreeNode:
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:28]: https://docs.python.org/3/library/dataclasses.html ?

File: data_structures/binary_tree/is_bst.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        raise ValueError(
+            "Each node should be type of TreeNode and data should be float."
+        )
+
+    def is_binary_search_tree_internal(
+        node: TreeNode | None, left_bound: float, right_bound: float
+    ) -> bool:
+        """
+        >>> is_binary_search_tree_internal(None)
+        True
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:98]: At least on test that returns `False`?

File: data_structures/binary_tree/is_bst.py (hunk 12 above)
Comment (alexpantyukhin) [reply data_structures/binary_tree/is_bst.py:31]: fixed

File: data_structures/binary_tree/is_bst.py (hunk 13 above)
Comment (alexpantyukhin) [reply data_structures/binary_tree/is_bst.py:109]: changed name

File: data_structures/binary_tree/is_bst.py (hunk 11 above)
Comment (alexpantyukhin) [reply data_structures/binary_tree/is_bst.py:22]: About runtime: We visit each node of tree only one time. So it's the linear time alg based on the number of nodes. About space: we don't add new data strcututre inside the function.

File: data_structures/binary_tree/is_bst.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        >>> is_binary_search_tree_recursive_check(TreeNode(data=1), 10, 20)
+        False
+        """
+
+        if node is None:
+            return True
+
+        return (
+            (node.data > left_bound)
+            and (node.data < right_bound)
```
Comment (cclauss) [inline data_structures/binary_tree/is_bst.py:119]:
```suggestion
            left_bound < node.data < right_bound
```

**Review by cclauss — APPROVED** [review]

**Review by CaedenPH — APPROVED** [review]

**Pull request merged**